src/react.tsx

```diff
@@ -6,6 +6,29 @@ import { equals, getDisplayName, keys, mapValues, some } from './utils'
 
 export type Diff<T, U> = Pick<T, Exclude<keyof T, keyof U>>
 
+class StoreSnapshotWrapper<StoreState extends object> implements Store<StoreState> {
+  constructor(
+    private snapshot: StoreSnapshot<StoreState>,
+    private callbackOnGetsubscriptions: (key: keyof StoreState) => void
+  ) { }
+  get<K extends keyof StoreState>(key: K) {
+    this.callbackOnGetsubscriptions(key)
+    return this.snapshot.get(key)
+  }
+  set<K extends keyof StoreState>(key: K) {
+    return this.snapshot.set(key)
+  }
+  on<K extends keyof StoreState>(key: K) {
+    return this.snapshot.on(key)
+  }
+  onAll() {
+    return this.snapshot.onAll()
+  }
+  getState() {
+    return this.snapshot.getState()
```

Comment:
Don't forget to call `callbackOnGetsubscriptions` here. We might want to split `callbackOnGetsubscriptions` into two callbacks - `onGet` and `onGetAll`. When the latter is called, we might dispose all existing `this.subscriptions` and replace them with a single `.onAll` subscription (which would be equivalent to what we have today).

So in the average case where a user sticks to `.get`, we get a significant speedup. If a user calls `.getState`, that de-optimizes to the pathological case.